### PR TITLE
Adding files as in latest MR 518 in umami on GitLab

### PR DESCRIPTION
**Pull request by jobirk**
## Summary

This pull request introduces the following changes

* adds the code from `umami/plotting` from the umami repo on GitLab (the version used it the one from [MR 518](https://gitlab.cern.ch/atlas-flavor-tagging-tools/algorithms/umami/-/merge_requests/518))
* adding the examples from `umami/examples/plotting` which demonstrate the puma package
* adding the unit tests from `umami/tests/unit/plotting`

Relates to the following issues

* #1

### CODE REVIEW

File: examples/plot_pie.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    draw_errors=False,
+    vertical_split=True,
+    plot_pie=True,
+    pie_colours=None,
+    pie_labels=HadrTruthLabel_labels,
+)
+
+plot_pie.add(histogram(df["HadronConeExclTruthLabelID"]))
+plot_pie.draw()
+plot_pie.savefig("pie_chart_HadronConeExclTruthLabelID.png")
```
Comment (manuguth) [inline examples/plot_pie.py:32]:
```suggestion
plot_pie.savefig("pie_chart_HadronConeExclTruthLabelID.png", transparent=False)
```

File: examples/plot_pt_vs_eff.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+logger.info("Plotting bkg rejection for inclusive efficiency as a function of pt.")
+# You can choose between different modes: "sig_eff", "bkg_eff", "sig_rej", "bkg_rej"
+plot_bkg_rej = var_vs_eff_plot(
+    mode="bkg_rej",
+    ylabel="Light-flavour jets rejection",
+    xlabel=r"$p_{T}$ [GeV]",
+    logy=False,
+    atlas_second_tag=(
+        "$\\sqrt{s}=13$ TeV, PFlow jets, \n$t\\bar{t}$ test sample, $f_{c}=0.018$"
```
Comment (manuguth) [inline examples/plot_pt_vs_eff.py:101]:
```suggestion
        "$\\sqrt{s}=13$ TeV, dummy jets, \n$t\\bar{t}$ test sample, $f_{c}=0.018$"
```

File: examples/plot_pt_vs_eff.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+plot_bkg_rej.draw()
+plot_bkg_rej.savefig("pt_light_rej.png")
+
+plot_sig_eff = var_vs_eff_plot(
+    mode="sig_eff",
+    ylabel="$b$-jets efficiency",
+    xlabel=r"$p_{T}$ [GeV]",
+    logy=False,
+    atlas_second_tag=(
+        "$\\sqrt{s}=13$ TeV, PFlow jets, \n$t\\bar{t}$ test sample, $f_{c}=0.018$"
```
Comment (manuguth) [inline examples/plot_pt_vs_eff.py:117]:
```suggestion
        "$\\sqrt{s}=13$ TeV, dummy jets, \n$t\\bar{t}$ test sample, $f_{c}=0.018$"
```

File: examples/plot_pt_vs_eff.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+plot_sig_eff.add(dips_light)
+
+plot_sig_eff.atlas_second_tag += "\nInclusive $\\epsilon_b=70%%$"
+
+# If you want to inverse the discriminant cut you can enable it via
+# plot_sig_eff.set_inverse_cut()
+plot_sig_eff.draw()
+# Drawing a hline indicating inclusive efficiency
+plot_sig_eff.draw_hline(0.7)
+plot_sig_eff.savefig("pt_b_eff.png")
```
Comment (manuguth) [inline examples/plot_pt_vs_eff.py:131]:
```suggestion
plot_sig_eff.savefig("pt_b_eff.png", transparent=False)
```

File: examples/plot_pt_vs_eff.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
+"""Produce pT vs efficiency plot from tagger output and labels."""
+import numpy as np
+
+from puma.utils.logging import logger
+from puma import var_vs_eff, var_vs_eff_plot
+from puma.utils import get_dummy_2_taggers
+
+
+# define a small function to calculate discriminant
+def disc_fct(a: np.ndarray) -> np.ndarray:
```
Comment (manuguth) [inline examples/plot_pt_vs_eff.py:9]: let's add here `fc` as an argument

File: examples/plot_rocs.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+from puma import roc, roc_plot
+
+# The line below generates dummy data which is similar to a NN output
+df = get_dummy_2_taggers()
+
+logger.info("caclulate tagger discriminants")
+
+
+# define a small function to calculate discriminant
+def disc_fct(a: np.ndarray) -> np.ndarray:
```
Comment (manuguth) [inline examples/plot_rocs.py:15]: also here to be consistent, we could add `fc` as argument

File: examples/plot_rocs.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+# n_test = 10_000
+
+# here the plotting of the roc starts
+logger.info("Plotting ROC curves.")
+plot_roc = roc_plot(
+    n_ratio_panels=2,
+    ylabel="background rejection",
+    xlabel="b-jets efficiency",
+    atlas_second_tag=(
+        "$\\sqrt{s}=13$ TeV, PFlow jets, \n$t\\bar{t}$ test sample, $f_{c}=0.018$"
```
Comment (manuguth) [inline examples/plot_rocs.py:80]:
```suggestion
        "$\\sqrt{s}=13$ TeV, dummy jets, \n$t\\bar{t}$ test sample, $f_{c}=0.018$"
```

File: examples/plot_rocs.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+plot_roc.set_ratio_class(2, "cjets", label="c-jets ratio")
+# if you want to swap the ratios just uncomment the following 2 lines
+# plot_roc.set_ratio_class(2, "ujets", label="light-flavour jets ratio")
+# plot_roc.set_ratio_class(1, "cjets", label="c-jets ratio")
+plot_roc.set_leg_rej_labels("ujets", "light-flavour jets rejection")
+plot_roc.set_leg_rej_labels("cjets", "c-jets rejection")
+
+
+plot_roc.draw()
+plot_roc.savefig("roc.png")
```
Comment (manuguth) [inline examples/plot_rocs.py:136]:
```suggestion
plot_roc.savefig("roc.png", transparent=False)
```

File: puma/plot_base.py [hunk 9]
```diff
@@ -3,18 +3,19 @@
...
+from puma.utils.logging import logger
 
 atlasify.LINE_SPACING = 1.3  # overwrite the default, which is 1.2
 
 
 # TODO: enable `kw_only` when switching to Python 3.10
 # @dataclass(kw_only=True)
 @dataclass
-class PlotLineObject:
+class plot_line_object:
```
Comment (manuguth) [inline puma/plot_base.py:17]: let's here directly keep the new naming, also below

File: puma/utils/__init__.py [hunk 10]
```diff
@@ -1,16 +1,337 @@
...
 
+import numpy as np
+import pandas as pd
+from scipy.special import softmax
 
-def set_xaxis_ticklabels_invisible(axis):
+from puma.utils.logging import logger
+
+
+def translate_kwargs(kwargs):
```
Comment (manuguth) [inline puma/utils/__init__.py:13]: don't think we still need this?

**Review by manuguth — CHANGES_REQUESTED** [review]

File: examples/plot_basic_histogram.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
+"""Example of histogram plot that deviates from umami default plots"""
```
[the remark this answers is not in the document]
Comment (jobirk) [reply examples/plot_basic_histogram.py]: yep, is changed

File: puma/plot_base.py (hunk 9 above)
Comment (jobirk) [reply puma/plot_base.py]: yeah I changed it back for `PlotLineObject` and `PlotObject`.

Do you think we should also already update `histogram_plot`, `histogram`, `roc_plot`, ... or should we move this in another PR?

Comment (manuguth) [reply puma/plot_base.py]: we can move this to another MR

File: examples/plot_rocs.py (hunk 6 above)
Comment (jobirk) [reply examples/plot_rocs.py]: I added it with `0.018` as default value

**Review by manuguth — APPROVED** [review]

**Pull request merged**